// @electron-forge/core/dist/api/init.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_path_1 = __importDefault(require("node:path"));
const core_utils_1 = require("@electron-forge/core-utils");
const chalk_1 = __importDefault(require("chalk"));
const debug_1 = __importDefault(require("debug"));
const listr2_1 = require("listr2");
const semver_1 = __importDefault(require("semver"));
const install_dependencies_1 = require("../util/install-dependencies");
const read_package_json_1 = require("../util/read-package-json");
const find_template_1 = require("./init-scripts/find-template");
const init_directory_1 = require("./init-scripts/init-directory");
const init_git_1 = require("./init-scripts/init-git");
const init_link_1 = require("./init-scripts/init-link");
const init_npm_1 = require("./init-scripts/init-npm");
const d = (0, debug_1.default)('electron-forge:init');
async function validateTemplate(template, templateModule) {
    if (!templateModule.requiredForgeVersion) {
        throw new Error(`Cannot use a template (${template}) with this version of Electron Forge, as it does not specify its required Forge version.`);
    }
    const forgeVersion = (await (0, read_package_json_1.readRawPackageJson)(node_path_1.default.join(__dirname, '..', '..'))).version;
    if (!semver_1.default.satisfies(forgeVersion, templateModule.requiredForgeVersion)) {
        throw new Error(`Template (${template}) is not compatible with this version of Electron Forge (${forgeVersion}), it requires ${templateModule.requiredForgeVersion}`);
    }
}
exports.default = async ({ dir = process.cwd(), interactive = false, copyCIFiles = false, force = false, template = 'base', skipGit = false, electronVersion = 'latest', }) => {
    d(`Initializing in: ${dir}`);
    const runner = new listr2_1.Listr([
        {
            title: `Resolving package manager`,
            task: async (ctx, task) => {
                ctx.pm = await (0, core_utils_1.resolvePackageManager)();
                task.title = `Resolving package manager: ${chalk_1.default.cyan(ctx.pm.executable)} v${ctx.pm.version}`;
            },
        },
        {
            title: `Resolving template: ${chalk_1.default.cyan(template)}`,
            task: async (ctx, task) => {
                const tmpl = await (0, find_template_1.findTemplate)(template);
                ctx.templateModule = tmpl.template;
                task.output = `Using ${chalk_1.default.green(tmpl.name)} (${tmpl.type} module)`;
            },
            rendererOptions: { persistentOutput: true },
        },
        {
            title: `Resolving Electron version: ${chalk_1.default.cyan(electronVersion)}`,
            task: async (ctx, task) => {
                if (electronVersion === 'latest' ||
                    electronVersion === 'beta' ||
                    electronVersion === 'nightly') {
                    task.output = `Using Electron version tag: ${chalk_1.default.cyan(electronVersion)}`;
                    ctx.parsedElectronVersion = electronVersion;
                }
                else {
                    // semver.clean allows us to accept `v` versions and trims whitespace
                    const maybeVersion = semver_1.default.clean(electronVersion);
                    if (maybeVersion) {
                        task.output = `Using Electron version: ${chalk_1.default.cyan(maybeVersion)}`;
                        ctx.parsedElectronVersion = maybeVersion;
                    }
                    else {
                        throw new Error(`Invalid Electron version: ${electronVersion}. Must be a valid semver version or one of 'latest', 'beta', or 'nightly'.`);
                    }
                }
            },
            rendererOptions: { persistentOutput: true },
        },
        {
            title: 'Initializing directory',
            task: async (_, task) => {
                await (0, init_directory_1.initDirectory)(dir, task, force);
            },
            rendererOptions: { persistentOutput: true },
        },
        {
            title: 'Initializing git repository',
            enabled: !skipGit,
            task: async () => {
                await (0, init_git_1.initGit)(dir);
            },
        },
        {
            title: 'Preparing template',
            task: async ({ templateModule }) => {
                await validateTemplate(template, templateModule);
            },
        },
        {
            title: `Initializing template`,
            task: async ({ templateModule }, task) => {
                if (typeof templateModule.initializeTemplate === 'function') {
                    const tasks = await templateModule.initializeTemplate(dir, {
                        copyCIFiles,
                        force,
                    });
                    if (tasks) {
                        return task.newListr(tasks, { concurrent: false });
                    }
                }
            },
        },
        {
            title: 'Installing template dependencies',
            task: async ({ templateModule }, task) => {
                return task.newListr([
                    {
                        title: 'Installing production dependencies',
                        task: async ({ pm }, task) => {
                            d('installing dependencies');
                            if (templateModule.dependencies?.length) {
                                task.output = `${pm.executable} ${pm.install} ${pm.dev} ${templateModule.dependencies.join(' ')}`;
                            }
                            return await (0, install_dependencies_1.installDependencies)(pm, dir, templateModule.dependencies || [], install_dependencies_1.DepType.PROD, install_dependencies_1.DepVersionRestriction.RANGE);
                        },
                        exitOnError: false,
                    },
                    {
                        title: 'Installing development dependencies',
                        task: async ({ pm }, task) => {
                            d('installing devDependencies');
                            if (templateModule.devDependencies?.length) {
                                task.output = `${pm.executable} ${pm.install} ${pm.dev} ${templateModule.devDependencies.join(' ')}`;
                            }
                            await (0, install_dependencies_1.installDependencies)(pm, dir, templateModule.devDependencies || [], install_dependencies_1.DepType.DEV);
                        },
                        exitOnError: false,
                    },
                    {
                        title: 'Finalizing dependencies',
                        task: async (ctx, task) => {
                            return task.newListr([
                                {
                                    title: 'Installing common dependencies',
                                    task: async ({ pm }, task) => {
                                        await (0, init_npm_1.initNPM)(pm, dir, ctx.parsedElectronVersion, task);
                                    },
                                    exitOnError: false,
                                },
                                {
                                    title: 'Linking Forge dependencies to local build',
                                    enabled: !!process.env.LINK_FORGE_DEPENDENCIES_ON_INIT,
                                    task: async ({ pm }, task) => {
                                        await (0, init_link_1.initLink)(pm, dir, task);
                                    },
                                    exitOnError: true,
                                },
                            ]);
                        },
                    },
                ], {
                    concurrent: false,
                });
            },
        },
    ], {
        concurrent: false,
        silentRendererCondition: !interactive,
        fallbackRendererCondition: Boolean(process.env.DEBUG) || Boolean(process.env.CI),
    });
    await runner.run();
};
